### DISCUSSION TITLE
Roadmap brainstorming

### DISCUSSION THREAD

**Discussion by ruiconti** [t10d/kingdom-python-server#1] [Ideas]
# Brainstorming

The goal here is to start discussion and collect actions that need to be done so we can properly priorize.

Since this is about going public, it makes sense to split tasks in 3 chunks:

## Repository management

- [ ] ~Proper documentation using sphinx or similars~
- [ ] ~Make it ready to be used w/ cookiecutter~

## Chore

- [x] ~Make tests work again~
- [ ] ~Adjustments and improvements on email adapter~
- [ ] Adjustments on CI/CD to make it work again
- [ ] Adjustments on serverless entrypoint to be agnostic, properly name things to AWS
- [ ] ~Address TODOs~

## Refactors

- [ ] ~Stricter `mypy` analysis~
- [ ] ~Inclusion of Interactor into auth module to enable proper dependency injection and cleaner decoupled interface between entrypoint and service layer.~
- [ ] ~Implement dockercompose scripts to start-up application and its dependencies~
- [ ] ~Re-implement integration tests w/ dockercompose instancing MailDog + Postgres~
- [ ] The permission domain model: the current model does not follows strictly either RBAC or ABAC, even that it reflects some proposals from both. We have a model with a optional relationship between a User and a Role, but the need for this being mandatory is start to show up and some specific Roles will be attached to business rules (e.g. linking a User to an Account), if having Role as a mandatory relationship for User we can remove the is_conditional attribute from Permission and check the access limit through the Role. I see this as a future refactor that should not block this PR.

## Application features

- [ ] Compliance to TLS
- [ ] Interoperability & translation of GraphQL queries to Postgres queries
- [ ] Compile typed code into C extensions using [mypyc](https://github.com/mypyc/mypyc)
- [ ] Proper linting
- [ ] Port every IO dependency to its async sibling
- [ ] Implement a toy aggragate example that demonstrate MessageBus behaviors
- [ ] Perform stress tests to evaluate performance under a simulated production environment
- [ ] Decouple aggregates into microservices w/ individual deployments (w/ docker-compose)
- [ ] Proper environment management
- [ ] HTTP session management
- [ ] RESTful entrypoint
    - [ ] Adherence to RFC and API guidelines in proper headers usage
- [ ] Improve logging & monitoring facilities
- [ ] Improve configuration management to make it easier to extend microservices & adapters
- [ ] Support background workers using celery queues + docker-compose
- [ ] Support background workers using GCP functions

**Comment by andreyrcdias** [comment]
An improvement that could be made is the better division of dependencies by environment. I saw that we already have something done in poetry, but I believe that we can make improvements to make it more dry, which is dev, test and production.

In the future we can also create a page on Github Pages similar to [this one](https://drivendata.github.io/cookiecutter-data-science/#cookiecutter-data-science), explaining possible benefits and usage

**Reply by ruiconti** [reply]
@andreyrcdias I agree with you. There is virtually nothing done here to split those environments and I've added this topic so we can properly priorize.

**Comment by ruiconti** [comment]
First Milestone [has been published](https://github.com/t10d/kingdom-python-server/milestone/1) :-)

Tasks that are related to it were crossed in this discussion.
